scripts/abins/instruments/directinstrument.py

```diff
@@ -0,0 +1,156 @@
+            # as this would omit contributions from the range -10 < θ < 10
+            min_angle = 0
+
+        q_limits = np.empty((2, energy.size))
+        q_limits[0, :] = np.sqrt(self.calculate_q_powder(input_data=energy, angle=min_angle))
+        q_limits[1, :] = np.sqrt(self.calculate_q_powder(input_data=energy, angle=max_angle))
+        return q_limits
+
+    def get_q_bounds(self, pad: float = 0.05) -> Tuple[float, float]:
+        """Calculate appropriate q range for instrument
+
+        Coarsely sample the outer E-Q lines to determine plotting extent
+
+        Args:
+            pad: Additional fraction of maximum value to use as padding
+
+        """
+        energy_samples = np.linspace(0, self.get_incident_energy(), 10)
+        q_values = self.get_abs_q_limits(energy_samples).flatten()
+        q_min, q_max = np.min(q_values), np.max(q_values)
+        if q_min - (pad * q_max) < 0:
+            q_min = 0
+        return (q_min, q_max * (1 + pad))
+
+    def calculate_q_powder(self, *, input_data=None, angle=None):
+        """
+        Returns momentum transfer Q^2 corresponding to frequency series
+
+        using cosine rule Q^2 = a^2 + b^2 - 2 ab cos(theta)
+        where a, b are k_i and k_f
+
+        Calculation is restricted to the region E < E_i
+
+        :param input_data: list with frequencies for the given k-point.
+        :param angle: scattering angle in degrees
+        :type array-like:
+        """
+
+        # Set momentum transfer to zero for energy transitions larger than the
+        # incident energy of neutrons and calculate momentum transfer according
+        # to momentum conservation principle for energy transitions smaller than
+        # the incident energy of neutrons.
+
+        k2_f = np.zeros_like(input_data)
+        k2_i = np.zeros_like(input_data)
+
+        e_init = self.get_incident_energy()
+
+        conservation_indx = e_init > input_data
+
+        k2_f[conservation_indx] = (e_init - input_data[conservation_indx]) * WAVENUMBER_TO_INVERSE_A
+        k2_i[conservation_indx] = e_init * WAVENUMBER_TO_INVERSE_A
+
+        cos_angle = math.cos(math.radians(angle))
+        result = k2_i + k2_f - 2 * cos_angle * (k2_i * k2_f) ** 0.5
+
+        return result
+
+    def convolve_with_resolution_function(self, frequencies=None, bins=None, s_dft=None, scheme='auto'):
+        """
+        Convolve discrete frequency spectrum with energy-dependent resolution function for a 2D instrument.
+
+        :param frequencies: DFT frequencies for which resolution function should be calculated (frequencies in cm-1)
+        :param s_dft:  discrete S calculated directly from DFT
+        """
+
+        if scheme == 'auto':
+            scheme = 'interpolate'
+
+        if scheme in prebin_required_schemes:
+            s_dft, _ = np.histogram(frequencies, bins=bins, weights=s_dft, density=False)
+            frequencies = (bins[1:] + bins[:-1]) / 2
```

Comment:
Are these the frequencies as point data?

Reply:
They are frequencies corresponding to the weights. This can be bin centres of pre-binned data, or a full set of frequencies and weights for individual modes. (Which might be automatically binned here if necessary.)

Reply:
what would be better name? frequencies_of_weights?

Reply:
`frequencies` is used pretty consistently around Abins for this data, I think calling it anything else here would be more confusing.

Reply:
Perhaps it would help to complete this docstring with more information about the _output_, clarifying that it will always be binned?

Reply:
How about:
- points_freq renamed to bin_centres
- docstring becomes
```
        """
        Convolve discrete frequency spectrum with energy-dependent resolution function for a 2D instrument.

        :param frequencies: DFT frequencies for which resolution function should be calculated (frequencies in cm-1)
        :param s_dft:  discrete S calculated directly from DFT

        :returns:
            (bin_centres, broadened_spectrum):
            Resolution functions always create binned data on a regular mesh.
            ``bin_centres`` and ``broadened_spectrum`` are a consistent set of
            mesh frequencies and corresponding broadened intensity.
        """